include/deal.II/lac/constraint_matrix.templates.h

```diff
@@ -639,15 +651,15 @@ distribute_local_to_global (const Vector<LocalType>      &local_vector,
         // constrained. if not, just write the entry
         // into the vector. otherwise, need to resolve
         // the constraint
-        if (is_constrained(local_dof_indices[i]) == false)
+        if (is_constrained(local_dof_indices_col[i]) == false)
           {
-            global_vector(local_dof_indices[i]) += local_vector(i);
+            global_vector(local_dof_indices_col[i]) += local_vector(i);
```

Comment:
At least for the code in line 647 (new), the situation is pretty clear that it must be `local_dof_indices_row` because the entries belong to the right hand. On a second thought, I think that the assembly process must only add the `local_vector` part if we are on the matrix diagonal (I would add a bool variable `diagonal` for that purpose to this function). The inhomogeneous part, of course, will need to take different row and col arguments. The same holds for the code in line 655/657 (new) which is basically the same spot.

Now, coming to the matrix part, I admit that the code is a bit involved, so let us take the two spots down separately. They are both correct, so the only place to fix is the above (together with ensuring that the vector part is only taken into account once when adding two or more blocks).